humility-core/src/arch.rs

```diff
@@ -2,339 +2,176 @@
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
+use crate::arch_arm::ARMRegister;
+use crate::arch_rv::RVRegister;
 use anyhow::{bail, Result};
-use num_traits::ToPrimitive;
-use std::collections::{BTreeMap, HashMap};
+use capstone::arch::arm;
+use capstone::arch::riscv;
+use capstone::arch::ArchOperand;
+use num_traits::cast::ToPrimitive;
+use num_traits::FromPrimitive;
 
 #[derive(Copy, Clone, Debug)]
-pub struct ARMRegisterField {
+pub struct RegisterField {
     pub highbit: u16,
     pub lowbit: u16,
     pub name: &'static str,
 }
 
-impl ARMRegisterField {
-    fn field(highbit: u16, lowbit: u16, name: &'static str) -> Self {
+impl RegisterField {
+    pub fn field(highbit: u16, lowbit: u16, name: &'static str) -> Self {
         Self { highbit, lowbit, name }
     }
-    fn bit(bit: u16, name: &'static str) -> Self {
+    pub fn bit(bit: u16, name: &'static str) -> Self {
         Self { highbit: bit, lowbit: bit, name }
     }
 }
-
 #[allow(non_camel_case_types)]
-#[derive(
-    Copy,
-    Clone,
-    Debug,
-    Hash,
-    FromPrimitive,
-    ToPrimitive,
-    PartialEq,
-    Eq,
-    Ord,
-    PartialOrd,
-)]
-///
-/// The definition of an ARM register, as encoded in the Debug Core Register
-/// Selector Register (DCRSR); see (e.g.) C1.6.3 in the ARM v7-M Architecture
-/// Reference Manual.
-///
-pub enum ARMRegister {
-    R0 = 0,
-    R1,
-    R2,
-    R3,
-    R4,
-    R5,
-    R6,
-    R7,
-    R8,
-    R9,
-    R10,
-    R11,
-    R12,
-    SP = 0b000_1101,
-    LR = 0b000_1110,
-    PC = 0b000_1111,
-    PSR = 0b001_0000,
-    MSP = 0b001_0001,
-    PSP = 0b001_0010,
-    SPR = 0b001_0100,
-    FPSCR = 0b010_0001,
-    S0 = 0b100_0000,
-    S1,
-    S2,
-    S3,
-    S4,
-    S5,
-    S6,
-    S7,
-    S8,
-    S9,
-    S10,
-    S11,
-    S12,
-    S13,
-    S14,
-    S15,
-    S16,
-    S17,
-    S18,
-    S19,
-    S20,
-    S21,
-    S22,
-    S23,
-    S24,
-    S25,
-    S26,
-    S27,
-    S28,
-    S29,
-    S30,
-    S31,
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, Ord, PartialOrd)]
+pub enum Register {
+    Arm(ARMRegister),
+    RiscV(RVRegister),
 }
 
-impl ARMRegister {
+impl Register {
     pub fn is_general_purpose(&self) -> bool {
-        matches!(
-            self,
-            ARMRegister::R0
-                | ARMRegister::R1
-                | ARMRegister::R2
-                | ARMRegister::R3
-                | ARMRegister::R4
-                | ARMRegister::R5
-                | ARMRegister::R6
-                | ARMRegister::R7
-                | ARMRegister::R8
-                | ARMRegister::R9
-                | ARMRegister::R10
-                | ARMRegister::R11
-                | ARMRegister::R12
-                | ARMRegister::SP
-                | ARMRegister::PC
-                | ARMRegister::LR
-        )
+        match self {
+            Register::Arm(reg) => reg.is_general_purpose(),
+            Register::RiscV(reg) => reg.is_general_purpose(),
```

Comment:
Yea, the current style is a bit tedious.  I do not know of a good way to access the underlying `reg` without matching each case.  I just chose this as it matches the other functions where the call is not identical.

For 64bit though, I don't think any changes will be needed for `Register` or `RVRegister`.

Reply:
I tried to use traits before, but ran into lots of problems with traits not actually being types or structs.  It would require restructuring all of the code that needs registers.  I think the way around this is to add a `RegisterTrait` trait which outlines all of the functions, then within `Register` use this trait to generically call the `RegisterTrait` functions, regardless of the enum variant.  

I did not do this as it was a lot more complicated, and doesn't really help clean it up that much.  However, if you think this would be better than just enumerating the 2 architectures, I can make the change.

Reply:
It might not be worth it especially considering we might not have to worry about more architectures at least in the short term. Thanks for checking this Drew!